### ISSUE TITLE
Implement ability to create custom dimensions from metric tiles

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#7601]
## Feature Description

A metric tile dependent on custom dimensions should use a new wrapper/utility that goes on to create the custom dimension(s) it depends on.

See:
1. [Relevant section in the design doc](https://docs.google.com/document/d/1htqStvvyXq_eg5GsHSbdBBgm4Tu_Z6KXsIOBfjWQ5oM/edit?resourcekey=0-Px-yDiCQOTz30DBKllhU3g#heading=h.q3f4ph3955po).
2. [Figma designs for error states](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?type=design&node-id=6285-85048&mode=design&t=jVtKY6g6XWmpvybS-4).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- A new wrapper component should be created for metric tiles that depend on custom dimensions that should act like the following:
  - If the required custom dimensions are available:
    - The wrapper should let the metric tile make the appropriate report request and display the report accordingly.
    - If the report request fails with an `INVALID_ARGUMENT` `error.status`, the locally persisted list of available custom dimensions might not be up to date and that should be updated. This update should be triggered only once so that multiple metric tiles do not attempt to update this at the same time.
  - If the required custom dimensions are **_not_** available:
    - The metric tile should show a loading state if the custom dimensions are being created in the background.
    - The metric tile should show an error state if the custom dimension creation for the current tile fails:
      - If the error occurs due to the user not having sufficient permissions for the GA property, the metric tile should show an error state for “Insufficient permissions” according to the Figma designs:
        - Heading: `Insufficient permissions`
        - Text: `You’ll need to contact your administrator. Learn more (link) <br> Permissions updated? Retry (link)`
        - Clicking on `Learn more` should take the user to the Site Kit support URL with an `error_id` of `analytics-4_insufficient_permissions`.
        - Clicking on `Retry` should attempt to create the custom dimensions again.
        - Screenshot for reference:
        ![image](https://github.com/google/site-kit-wp/assets/20284937/46cbc2b4-4393-4433-bbbf-853b03c4b9c3)
      - If the error occurs due to something generic, the metric tile should show an error state with a generic error message according to the Figma designs:
        - Heading: `Analytics update failed`.
        - Text: `Retry didn't work? Get help (link)`
        - CTA: `Retry`
        - Clicking on `Get help` should take the user to the Site Kit support URL with an `error_id` of the encountered error code.
        - Clicking on `Retry` should attempt to create the custom dimensions again.
        - Screenshot for reference:
        ![image](https://github.com/google/site-kit-wp/assets/20284937/44b95e10-9890-4a54-9316-11d6a4493366)
    - Otherwise, it should be checked if the user has the Analytics “*edit*” scope.
      - If the user doesn’t have the Analytics “*edit*” scope:
        - The metric tile should show an error state according to the Figma designs:
          - Heading: `No data to show`
          - Text: `Update Analytics to track metric`
          - CTA: `Update`
          - Clicking on `Update` should direct the user to the OAuth flow to grant the Analytics “*edit*” scope in such a way so that Site Kit automatically creates the custom dimensions after they return. This behaviour should already be partially implemented in [#7599](https://github.com/google/site-kit-wp/issues/7599) and the same should be adopted here as well.
      - If the user has the Analytics “*edit*” scope (or they just granted it) and Site Kit doesn't automatically create the custom dimensions, the metric tile should show an error state according to the Figma designs:
          - Heading: `No data to show`
          - Text: `Update Analytics to track metric`
          - CTA: `Update`
          - Clicking on `Update` should trigger the creation of custom dimensions.
          - Screenshot for reference:
          ![image](https://github.com/google/site-kit-wp/assets/20284937/ff814007-0382-463a-bbc2-e0e8a06de147)
  - Until an error scenario, a gathering/zero data state, or a report is displayed, the metric tile should show a consistent loading state. It should be ensured that there are no flickers between different states.
- This wrapper should be a drop-in component so that it automatically handles the creation of custom dimensions and relevant error scenarios. The issues implementing the metric tiles ([#7603](https://github.com/google/site-kit-wp/issues/7603), [#7605](https://github.com/google/site-kit-wp/issues/7605), [#7607](https://github.com/google/site-kit-wp/issues/7607)) should be able to use this wrapper effortlessly and not have to worry about creating the custom dimensions and handling errors.

## Implementation Brief

* Create `assets/js/components/KeyMetrics/CustomDimensionTileWrapper`
  * Add `CustomDimensionErrorActions.js`
    * Use this markup as starting point - https://github.com/google/site-kit-wp/blob/91ce4cd17f8d234c5baef2364b62cc641f532e6b/assets/js/components/ReportErrorActions.js#L118-L135
    * Include props to handle different variations of content and link/callback function by adapting different output of the referenced code. 
    * When learn more is passed it should render text and link/callback fn as defined in AC for `Insufficient permissions` tile, as well as prop for inline `retry` defined on the same tile
    * When get help is present it should render text and link defined in AC for `generic error`
    * Button callback function and label can be more dynamic to adapt per need when passed as a props
  * Add main error component `CustomDimensionError.js`
    * Use `assets/js/components/KeyMetrics/MetricTileError/index.js` as starting point
    * Move current `title` value to the variable so it can be adapted to different messages
    * Create callback functions to handle granting the `EDIT_SCOPE` permission (Example in [#7599](https://github.com/google/site-kit-wp/issues/7599)), and to create custom dimensions by dispatching `createCustomDimensions` action introduced in [#7598](https://github.com/google/site-kit-wp/issues/7598)
    * Use `CustomDimensionErrorActions` component for content output
    * You can use variables to determine which components will be rendered in `CustomDimensionErrorActions`, and which callback to use for links/button
    * User having/or not having `EDIT_SCOPE` should modify error output with content defined in design and AC for that scenario
    * In case of `hasInsufficientPermissionsError` title and content should match `Insufficient permissions` tile in AC
  * Add main tile wrapper component `index.js`
    * Include props to accept required custom dimensions and `reportOptions`
    * You can use `hasCustomDimensions` selector to determine if custom dimensions are available
    * Check for potential errors with `getErrorForSelector` on the report made by child component, using `reportOptions` prop
    * In `useMount` or `useEffect` hook, do early check if `error.status` is `INVALID_ARGUMENT` and if  `getAvailableCustomDimensions` is not fetching data. 
      * And In that case  dispatch `syncAvailableCustomDimensions`  action, which will in turn create missing custom dimensions and sync the `availableCustomDimensions` state. Just implement `isSyncing` selector, and `setIsSyncing` controller, in that store to lock `syncAvailableCustomDimensions` action by returning early if it `isSyncing`, to prevent multiple calls at the same time.
    * Check if metrics are being created, to induce the loading state. You can use `isFetchingCreateCustomDimension`, or perhaps `hasFinishedResolution` on the `getReport`. And to unify it with child component loading, you can add additional loading prop and combine it in this check 
    * On loading return `MetricTileLoader` component, with necessary wrappers to keep the styling, you can get them from `MetricTileNumeric` component for example.
    * On error that is not due to `INVALID_ARGUMENT`, return `CustomDimensionError` component
    * Otherwise children should be returned

Here is a summary of the component hierarchy for clarity:

1. Custom dimensions don't exist and are not being created: `CustomDimensionTileWrapper > CustomDimensionError component (with prompt to update Analytics)`.
2. Custom dimensions being created: `CustomDimensionTileWrapper > MetricTileLoader`.
3. Custom dimensions creation failed: `CustomDimensionTileWrapper > CustomDimensionError component (with relevant error)`.
4. Custom dimensions exist: `CustomDimensionTileWrapper > Child Tile component`.

### Storybook

* Add Storybook stories for the `CustomDimensionError` component including its various variants.

### Test Coverage

* Add JS tests for the wrapper component.
* Add VRT scenarios for added Storybook stories.

## QA Brief

- Set up Site Kit and connect the Analytics module.
- Enable the `newsKeyMetrics` feature flag.
- In the appearing KMW banner notification, choose to pick up your own metrics and select the "Top recent trending pages", "Top categories by pageviews", and "Most popular authors by pageviews" metrics to show in your dashboard.
- Upon saving your selection, you should be directed to the OAuth flow. Cancel the OAuth flow to return to the SK dashboard.
- In the selected metric tiles, verify that you see an error state with a prompt to update your Analytics property.
- Verify that clicking on "Update" makes the tiles go on a loading state, creates the custom dimensions in the background (you can observe network requests in developer tools) and the tiles are displayed without an error state.
  
#### Verify insufficient permissions error state

- Follow the above steps, but connect an Analytics property to Site Kit on which you do not have "Editor" or "Administrator" access.
- Upon clicking on "Update" in the error state, the tiles should go on a loading state and attempt to create the custom dimensions in the background.
- Since your account doesn't have at least an "editor" access to the GA property, verify that the request fails and you see an error tile with "Insufficient permissions". Verify that the "Learn more" and "Retry" links work as expected.

#### Verify generic error state

- Follow the first set of steps.
- Just before clicking the "Update" CTA in the error state, use the browser developer tools and simulate an "Offline" network status.
- Click on "Update" and verify that a different error state appears that says "Analytics update fails".
- Verify that the "Learn more" link and "Retry" CTA works as expected.

#### Verify scenario where custom dimension is deleted after creation

- Select the `Top categories by pageviews` metric tile to be visible in the KMW area.
- Complete custom dimension setup and wait for the report request to be successful.
-  Go to analytics.google.com -> Admin. Select your connected property and go to `Custom definitions`. Archive the 'googlesitekit_post_categories` custom dimension.
- In the SK dashboard, clear session storage and reload the page.
- In the network tab of developer tools, verify that the report request for this tile fails with a `badRequest` error reason. Once that fails, SK should make another request to sync the custom dimensions and then show the missing dimensions error state ("Update Analytics to track metric") in the tile.
7. Clicking on "Update" should create the custom dimension again and re-fetch the report request.

## Changelog entry

* Allow creation of custom Analytics dimensions via Key Metric tiles.

**Comment by techanvil** [comment]
Hey @nfmohit, this AC is well written but again, it's going into IB territory where it doesn't really need to.

- Arguably, the point about `syncCustomDimensions` could be written in a more behaviourally descriptive way (with a note in the IB to capture the implementation detail), but it's a minor point and we can skip that if you want.
- However, I think the points that describe the `autoSubmit` behaviour would be better off being removed from here, and rather the AC can refer to the fact this behaviour should already be partially implemented via #7599, the implementation side of which can be subsequently referred to in the IB.
- Relating to the previous point, when it comes to the behaviour upon returning from OAuth, there seems to be a bit of a crossover between this issue as described and 7599, as the AC in this one currently states that "the metric tile should trigger creation of the custom dimensions automatically", whereas this should already be taken care of via "the Key Metrics Widget area should create the custom dimensions automatically" in 7599, if I've understood things correctly. Does this indicate that 7599 should in fact specify that the individual _tiles_ should create the custom dimensions? Or, maybe that detail can be left to the IB while this issue can simply refer back to 7599 as mentioned above.
- A minor grammatical point, relating to the first two sub-points in the AC: The custom dimensions are dependencies, not dependants, of the tiles. So, rather than describing them as "the dependent custom dimensions", it would be more correct to say for example "the required custom dimensions".

**Comment by nfmohit** [comment]
Thank you for the kind review, @techanvil. I've addressed your feedback as the following:

> Arguably, the point about `syncCustomDimensions` could be written in a more behaviourally descriptive way (with a note in the IB to capture the implementation detail), but it's a minor point and we can skip that if you want.

This has been updated.

> However, I think the points that describe the `autoSubmit` behaviour would be better off being removed from here, and rather the AC can refer to the fact this behaviour should already be partially implemented via #7599, the implementation side of which can be subsequently referred to in the IB.

Updated as well.

> Relating to the previous point, when it comes to the behaviour upon returning from OAuth, there seems to be a bit of a crossover between this issue as described and 7599, as the AC in this one currently states that "the metric tile should trigger creation of the custom dimensions automatically", whereas this should already be taken care of via "the Key Metrics Widget area should create the custom dimensions automatically" in 7599, if I've understood things correctly. Does this indicate that 7599 should in fact specify that the individual _tiles_ should create the custom dimensions? Or, maybe that detail can be left to the IB while this issue can simply refer back to 7599 as mentioned above.

You're correct, this was a mistake. This should actually rely on the key metrics widget area for custom dimensions creation if Site Kit is supposed to create them automatically. If Sit Kit is not supposed to create them automatically, the error state CTA should trigger it. I've updated the section.

> A minor grammatical point, relating to the first two sub-points in the AC: The custom dimensions are dependencies, not dependants, of the tiles. So, rather than describing them as "the dependent custom dimensions", it would be more correct to say for example "the required custom dimensions".

Thank you for pointing out, updated.

Please let me know if this looks good, thank you!

**Comment by techanvil** [comment]
Thanks @nfmohit, that looks great!

AC LGTM :white_check_mark:

**Comment by aaemnnosttv** [comment]
Thanks @zutigrm! A few general points about the IB as it relates to the guidelines which I encourage you to review as a refresher.

- The IB as a whole is a bit verbose and could be simplified to read a bit more concisely. For example, we don't need to prescribe creating a folder explicitly, we can simply add this to the path of a new component we're describing. E.g. `components/KeyMetrics/CustomDimensionsMetricTileError/ErrorActions`
- The IB shouldn't duplicate text that is defined in the AC since this can get out of sync during definition.
- We want to keep the IB to be more of a summary of the approach, so it should be limited in the level of detail it includes to only that which is necessary. Things like names of internal functions and variables should be omitted as well as ternary operators or other specific code details. The IB should describe the approach rather than _how_ to arrive at some code that sounds like it has already been written. It's a bit hard to follow as-is.
- Regarding the code references, please always use GitHub "permalinks" to ensure they always reference the same source. The current links use the branch and line numbers only, which will change as the source updates making it easy to get out of sync.

---

> * Create a new `CustomDimensionsMetricTileError` folder

This directory probably isn't necessary as we can add the `ErrorActions` under the `CustomDimensionTileWrapper` since this component should only be relevant for metric tiles that depend on our custom dimensions.

> * Create new `ErrorActions.js` for component which will handle the error text and button actions

AFAIU this would only be used for errors related to custom dimension actions, correct? If so, let's name this accordingly as I believe we'd still be using `ReportErrorActions` within the metric tile itself for handling report errors. Perhaps `CustomDimensionErrorActions`.

> * Dispatch `invalidateResolution` on `analytics-4` store for `getAvailableCustomDimensions`, so it can trigger re-fetch once, on first tile re-render

This may be insufficient as most resolvers have a condition to return early if there is already data in state for it which we will still likely want to keep/implement for consistency, but also to minimize the need for fetch mocks in tests. One way around this could be to have an additional flag in state that we can enable when we want to do this to bypass this guard in the resolver. Using a resolver for this is a good idea though because it solves the issue of ensuring it's only done once, regardless of how many custom dimensions error.

---

Question about the AC for @nfmohit 

> If the required custom dimensions are **_not_** available:
> * The metric tile should show a loading state if the custom dimensions are being created in the background.

Is this issue intended to implement the background creation mentioned here? If not, where is this being handled?

**Comment by zutigrm** [comment]
@aaemnnosttv Thank you for the feedback.

> Dispatch invalidateResolution on analytics-4 store for getAvailableCustomDimensions, so it can trigger re-fetch once, on first tile re-render
This may be insufficient as most resolvers have a condition to return early if there is already data in state for it which we will still likely want to keep/implement for consistency, but also to minimize the need for fetch mocks in tests. One way around this could be to have an additional flag in state that we can enable when we want to do this to bypass this guard in the resolver. Using a resolver for this is a good idea though because it solves the issue of ensuring it's only done once, regardless of how many custom dimensions error.

What about using the `receiveGetAvailableCustomDimensions` with value of `undefined` to reset the state, and then invalidate the resolver?

> Is this issue intended to implement the background creation mentioned here? If not, where is this being handled?

To chip in here, creation of metrics and it's actions is being added in #7598. `createCustomDimensions` will be used to create custom metrics, and until it is resolved, by my understanding it will be creating the metrics. That's why I suggested waiting for `isFetchingCreateCustomDimension` to determine if metrics are still being created.

**Comment by nfmohit** [comment]
> Question about the AC for @nfmohit
> 
> > If the required custom dimensions are **_not_** available:
> > 
> > * The metric tile should show a loading state if the custom dimensions are being created in the background.
> 
> Is this issue intended to implement the background creation mentioned here? If not, where is this being handled?

Hi @aaemnnosttv. Triggering the creation of custom dimensions is a part of [#7599](https://github.com/google/site-kit-wp/issues/7599). However, it is also notable that the error state introduced in this current issue will also trigger automatic creation of custom dimensions.

**Comment by aaemnnosttv** [comment]
> What about using the `receiveGetAvailableCustomDimensions` with value of `undefined` to reset the state, and then invalidate the resolver?

We generally don't allow receiving `undefined`, but even if we did this would be problematic because all of the widget tiles that depend on custom dimensions would "see" that their dimension isn't available and would attempt to create it, or show an error state that didn't relate to them if only a single dimension was missing for example. So we want to keep the current state and simply sync/refetch the available dimensions. This can probably be done using a kind of "lock" in state to set before we start, to ensure that it only happens once (if multiple attempts are invoked at the same time).

So the action that triggers the sync, could look something like this

- Check isSyncing
	- If `true`, return
- Set isSyncing
- Dispatch fetch
- Unset isSyncing
- (return response?) The fetch should receive the updated dimensions in state as part of its `reducerCallback` so the return might not be needed

> Hi @aaemnnosttv. Triggering the creation of custom dimensions is a part of [#7599](https://github.com/google/site-kit-wp/issues/7599). However, it is also notable that the error state introduced in this current issue will also trigger automatic creation of custom dimensions.

@nfmohit I may be misunderstanding you, but I don't think the error state would trigger _automatic_ creation of custom dimensions, that should be tied to the "Update" button no? Dimension creation should only be triggered automatically by the `autoSubmit` when selecting new metric tiles that depend on them, correct?

**Comment by zutigrm** [comment]
@aaemnnosttv Thank you for the feedback.

> @nfmohit I may be misunderstanding you, but I don't think the error state would trigger automatic creation of custom dimensions, that should be tied to the "Update" button no? Dimension creation should only be triggered automatically by the autoSubmit when selecting new metric tiles that depend on them, correct?

@nfmohit means on the part in AC, which is mentioning that on `INVALID_ARGUMENT`  error for custom dimensions, it should trigger re-creating of the missing required custom dimensions. For that purpose re-fetching in the IB is mentioned and the need for state locking

**Comment by nfmohit** [comment]
> > Hi @aaemnnosttv. Triggering the creation of custom dimensions is a part of [#7599](https://github.com/google/site-kit-wp/issues/7599). However, it is also notable that the error state introduced in this current issue will also trigger automatic creation of custom dimensions.
> 
> @nfmohit I may be misunderstanding you, but I don't think the error state would trigger _automatic_ creation of custom dimensions, that should be tied to the "Update" button no? Dimension creation should only be triggered automatically by the `autoSubmit` when selecting new metric tiles that depend on them, correct?

Hi @aaemnnosttv, apologies for the confusion. By _automatic creation of custom dimensions_, I meant the following error state:

![image](https://github.com/google/site-kit-wp/assets/20284937/9d249e00-c25e-40b2-840c-a332e2b08634)

When the user clicks on "Update", we should set `autoSubmit` to `true` and send the user to OAuth. In this case, the custom dimensions creation will be triggered by the dashboard upon the user's coming back.

Please let me know if that clarifies things. Thank you!

**Comment by aaemnnosttv** [comment]
Thanks for the iterations here @nfmohit!
IB ✅

**Comment by nfmohit** [comment]
While testing this implementation with the recently implemented "Top categories by pageviews" metric tile that uses the HOC introduced here, I noticed two bugs:

1. There is a slight flicker in this metric tile with the missing custom dimensions error state appearing when the page loads due to the `availableCustomDimensions` setting not being loaded yet.
2. There is an infinite loop for the sync available custom dimensions request when the HOC detects that any custom dimension it requires is no longer available.

I've created a follow-up PR #7712 that fixes the above issues. Sending this back to CR.

**Comment by nfmohit** [comment]
@aaemnnosttv and I have detected a new issue for which I have added another follow-up PR #7719.

We currently use an error.reason of `badRequest` to determine if a report request was made with unavailable custom dimensions, and use that to sync the list of available custom dimensions. `badRequest` is a very generic error and has a high potential to create an infinite loop. This PR solves that by checking for a specific error message that occurs due to invalid custom dimensions.

**Comment by aaemnnosttv** [comment]
Merged, thanks @nfmohit !

**Comment by wpdarren** [comment]
Still testing this.  At the generic error message tests. Will continue testing tomorrow. 

**Verified:**

- In the selected metric tiles, an error state with a prompt to update my Analytics property appears.
- Since your account doesn't have at least an "editor" access to the GA property, the request fails and you see an error tile with "Insufficient permissions".  The `Retry` links work as expected.  There is currently an issue with the `Learn more` link but its related to the service engineering side. 


<details>
 <summary>Screenshots</summary>

![image](https://github.com/google/site-kit-wp/assets/73545194/c2cce57b-2038-4343-96bd-3389ce07b11e)
![image](https://github.com/google/site-kit-wp/assets/73545194/6483756b-c91f-4e45-b1ca-3c5a95c70f9d)
![image](https://github.com/google/site-kit-wp/assets/73545194/6a944bf6-75e8-4485-ac95-0bd5168f25cb)
![image](https://github.com/google/site-kit-wp/assets/73545194/eb345585-f1b1-42a5-9ebf-5722f5fb6be8)
![image](https://github.com/google/site-kit-wp/assets/73545194/c0c9d2f1-1c43-4beb-b2ce-5b71583f98eb)
![image](https://github.com/google/site-kit-wp/assets/73545194/ad52520a-46b1-4e99-b843-433a8daa0b98)

</details>

**Comment by bethanylang** [comment]
@wpdarren Just chatted with @aaemnnosttv about this one and it seems like there are actually issues here on both the service side **and** the plugin side.

For the service, this link was added to the [Error States MVP Mapping (Get help) sheet](https://docs.google.com/spreadsheets/d/12x39d7j0GNqha0m7s4-Y3P5SwK9wwgxxxDnWp3svSKg/edit?resourcekey=0-IngulwDvPmI4SEq2g-f4rA#gid=0) as opposed to the [Documentation URLs (Learn more) sheet](https://docs.google.com/spreadsheets/d/19dgLTylSt6rbBDZi5yiIYuCbuHA75KLjZMgAwlwfYBg/edit#gid=1620222322). This was 100% on me as I had asked @jamesozzie to make this change. James, I'll follow up with an Asana task for you to get this added to Documentation URLs (Learn more) instead.

For the plugin, the URL that's being generated for the Learn more link is https://sitekit.withgoogle.com/support?error=The%20caller%20does%20not%20have%20permission, which is not the correct structure. Because there is no page at that URL, you're being directed to the fallback URL of https://sitekit.withgoogle.com/documentation/troubleshooting/using-troubleshooting-mode/, which is also not particularly helpful. I'll open up a new Asana task to get this fallback URL changed.

But the tl;dr here is that there is, in fact, an issue in the plugin in terms of how this URL is being served that will need to be investigated and fixed.

[Ongoing Slack thread about all of this here!](https://10up.slack.com/archives/CBKKQEBR9/p1698754396097929) I'm offline beginning tomorrow but Evan is aware of this issue and can help as needed. Thanks!

**Comment by aaemnnosttv** [comment]
> For the service, this link was added to the [Error States MVP Mapping (Get help) sheet](https://docs.google.com/spreadsheets/d/12x39d7j0GNqha0m7s4-Y3P5SwK9wwgxxxDnWp3svSKg/edit?resourcekey=0-IngulwDvPmI4SEq2g-f4rA#gid=0) as opposed to the [Documentation URLs (Learn more) sheet](https://docs.google.com/spreadsheets/d/19dgLTylSt6rbBDZi5yiIYuCbuHA75KLjZMgAwlwfYBg/edit#gid=1620222322). This was 100% on me as I had asked @jamesozzie to make this change. James, I'll follow up with an Asana task for you to get this added to Documentation URLs (Learn more) instead.

I just looked into this a bit more closely and so long as the intention is to use the same invalid permissions error that we have now, then the support URL was added to the correct sheet (as an error). The plugin needs to be updated to rewrite the error as we're doing for report errors to provide the proper `code` which will be used as the `error_id` which isn't currently in place.

https://github.com/google/site-kit-wp/blob/73868dca2ffdb31d3525780d96c669bfe8ba28e6/assets/js/components/ReportErrorActions.js#L78-L80

https://github.com/google/site-kit-wp/blob/73868dca2ffdb31d3525780d96c669bfe8ba28e6/assets/js/modules/analytics-4/components/key-metrics/InsufficientPermissionsError.js#L48-L50

> Can you please clarify what the correct URL structure should be, so that even if the correct URL isn't being served from the service, Darren can confirm that the fix is correct on the plugin side?

It should be 
```
https://site-kit-dev.appspot.com/support/?error_id=analytics-4_insufficient_permissions
```

Sending this back for a follow-up 👍

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@nfmohit an observation on the final section:  `Verify scenario where custom dimension is deleted after creation`
In the QAB it mentions:

> In the network tab of developer tools, verify that the report request for this tile fails with a badRequest error reason. Once that fails, SK should make another request to sync the custom dimensions and then show the missing dimensions error state ("Update Analytics to track metric") in the tile.

I see the `badRequest` error reason in the network tab but do not see the `Update Analytics to track metric` error in the tile. I cleared session storage and also cleared cookies and logged back in but no error message appears.  I just see the gathering state which is was it was after the initial oAuth. 

Any ideas why this could be? 

![image](https://github.com/google/site-kit-wp/assets/73545194/59968ddc-3a0a-43da-bdce-f38a780a8c01)
![image](https://github.com/google/site-kit-wp/assets/73545194/6afddafc-8c7f-4b96-be14-fabc68101919)

**Verified:**

- When following the QAB, I see an error state with a prompt to update your Analytics property and when clicking on "Update" it makes the tiles go on a loading state, creates the custom dimensions in the background and the tiles are displayed without an error state.
- Since my account doesn't have at least an "editor" access to the GA property, the request fails and I see an error tile with "Insufficient permissions".  The `Retry` and `Learn more`  links work as expected.  
- When following the QAB, a different error state appears that says `Analytics update failed` and the "Learn more" link and "Retry" CTA work as expected.

<details>
 <summary>Screenshots</summary>

![image](https://github.com/google/site-kit-wp/assets/73545194/c2cce57b-2038-4343-96bd-3389ce07b11e)
![image](https://github.com/google/site-kit-wp/assets/73545194/6483756b-c91f-4e45-b1ca-3c5a95c70f9d)
![image](https://github.com/google/site-kit-wp/assets/73545194/6a944bf6-75e8-4485-ac95-0bd5168f25cb)
![image](https://github.com/google/site-kit-wp/assets/73545194/eb345585-f1b1-42a5-9ebf-5722f5fb6be8)
![image](https://github.com/google/site-kit-wp/assets/73545194/c0c9d2f1-1c43-4beb-b2ce-5b71583f98eb)
![image](https://github.com/google/site-kit-wp/assets/73545194/ad52520a-46b1-4e99-b843-433a8daa0b98)
![image](https://github.com/google/site-kit-wp/assets/73545194/b441a1a6-b363-46e7-a537-b3e22b0d96ac)
![image](https://github.com/google/site-kit-wp/assets/73545194/e36c8371-dbee-48d4-89c7-257be759a8ed)
![image](https://github.com/google/site-kit-wp/assets/73545194/3773b27c-5998-4b94-a69d-79cda2a5d632)

</details>

**Comment by aaemnnosttv** [comment]
@wpdarren & @nfmohit from what I can see, the sync in response to the missing custom dimension error in the main report works as expected on the current `main`. The problem observed in the report above which is attempted to be worked around in @nfmohit 's PR above is specific to the case where we do not yet know if the custom dimension has data available or not (i.e is still gathering data) which makes it quite an edge case, although somewhat easy for us to encounter since we're going through the motions rather unrealistically quick. The report which fails with the expected error is the gathering data check, not the "main" report request that triggers the sync as defined in this issue. I still think we should address it but I don't feel that this should block this issue, as it's more about the effect of #7638.

Sending this back to QA as is in case there is anything else to check here, and I'll open a new issue about the remaining work for this to work properly with dimensions that are still gathering data.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

Please note that I went through the ACs and can confirm that all of the error states work as expected, other than the issue reported above about the gathering state. 

- I see an error state with a prompt to update your Analytics property when following the QAB instructions.
- Clicking on "Update" makes the tiles go on a loading state, creates the custom dimensions in the background. This was also observed in network requests in developer tools.  The tiles are displayed without an error state.

**Insufficient permissions error state**

- Upon clicking on "Update" in the error state, the tiles go on a loading state and attempt to create the custom dimensions in the background. Since my account doesn't have at least an "editor" access to the GA property, the request fails and you see an error tile with "Insufficient permissions". The "Learn more" and "Retry" links work as expected.

**Generic error state**

- The "Learn more" link and "Retry" CTA works as expected.

Custom dimension is deleted after creation

- In the network tab of developer tools, the report request for this tile fails with a `badRequest `error reason. 
- Once that fails, SK shows the missing dimensions error state `Update Analytics to track metric` in the tile.
- Clicking on "Update" creates the custom dimension again and re-fetch the report request.

<details>
 <summary>Screenshots</summary>

![image](https://github.com/google/site-kit-wp/assets/73545194/ee30f55d-fe1d-4e98-8cee-4139a1138b47)
![image](https://github.com/google/site-kit-wp/assets/73545194/6a8f12fa-2614-4077-a62a-ed3886950736)
![image](https://github.com/google/site-kit-wp/assets/73545194/871343f6-bd91-472c-aa8a-27d9b9d3d339)
![image](https://github.com/google/site-kit-wp/assets/73545194/794cf6e5-f9db-4fd0-8e89-700765828162)
![image](https://github.com/google/site-kit-wp/assets/73545194/e4107433-3937-46c9-9d74-3c04080e8ede)
![image](https://github.com/google/site-kit-wp/assets/73545194/e6c60445-2fff-4bdc-ad48-e53f004c5bfa)

</details>
